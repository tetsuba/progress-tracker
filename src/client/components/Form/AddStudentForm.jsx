import * as React from 'react'
import { Button, Col, Form } from 'react-bootstrap'
import { useInputChange } from '../../hooks/hooks'
import { useMutation } from '@apollo/react-hooks'
import { STUDENTS_QUERY } from '../../api/student/student.query'
import { ADD_NEW_STUDENT_MUTATION } from '../../api/student/student.mutation'

type Props = {
  toggleModal: () => void,
}

export default function AddStudentForm(props: Props) {
  const { toggleModal } = props
  const [addNewStudent] = useMutation(ADD_NEW_STUDENT_MUTATION, {
    refetchQueries: [{ query: STUDENTS_QUERY }],
  })
  const [inputs, setInputs] = useInputChange({
    firstName: '',
    lastName: '',
    DOB: '',
    teacherID: 'teacherId1234',
  })
  return (
    <Form
      id="AddStudentForm"
      onSubmit={(e) => {
        e.preventDefault()
        addNewStudent({ variables: { input: inputs } })
          .then(({ data }) => {
            toggleModal()
          })
          .catch((err) => console.log(err))
      }}
    >
      <Form.Row className="pt-3">
        <Form.Group as={Col} controlId="formGridFirstName">
          <Form.Control
            size="lg"
            placeholder="First Name"
            onChange={setInputs}
            value={inputs.firstName}
            name="firstName"
            required
          />
        </Form.Group>
      </Form.Row>
      <Form.Row className="pt-3">
        <Form.Group as={Col} controlId="formGridLastName">
          <Form.Control
            size="lg"
            placeholder="Last Name"
            onChange={setInputs}
            value={inputs.lastName}
            name="lastName"
            required
          />
        </Form.Group>
      </Form.Row>
      <Form.Row className="pt-3">
        <Form.Group as={Col} controlId="formGridDOB">
          <Form.Label>Date of birth</Form.Label>
          <Form.Control
            placeholder="DOB"
            onChange={setInputs}
            value={inputs.DOB}
            name="DOB"
            required
          />
        </Form.Group>
      </Form.Row>

      <Form.Row>
        <Button
          className="float-right"
          variant="secondary"
          onClick={toggleModal}
        >
          Close
        </Button>
        <Button className="float-right ml-2" variant="primary" type="submit">
          Submit
        </Button>
      </Form.Row>
    </Form>
  )
}
