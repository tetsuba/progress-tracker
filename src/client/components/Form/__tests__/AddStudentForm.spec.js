import { act } from 'react-dom/test-utils'
import {
  delay,
  graphRenderer,
  updateTextInput,
} from '../../../../test/testHelper'

import AddStudentForm from '../AddStudentForm'
import {
  addNewStudentMutationSuccess,
  getStudentsQuerySuccess,
} from '../../../../test/mockApi/student/studentMockMutation'

describe('<AddStudentForm>', () => {
  describe('@Render', () => {
    it('should render', async () => {
      let wrapper
      await act(async () => {
        wrapper = graphRenderer(AddStudentForm, [], {})
        await delay()
      })
      wrapper.update()
      expect(wrapper.find(AddStudentForm)).toMatchSnapshot()
    })
  })

  describe('@Events', () => {
    describe('onSubmit', () => {
      it('should call "toggleModal"', async () => {
        const props = {
          toggleModal: jest.fn(),
        }

        const wrapper = graphRenderer(
          AddStudentForm,
          [addNewStudentMutationSuccess, getStudentsQuerySuccess],
          props
        )

        act(() => {
          updateTextInput(wrapper, 'firstName', 'John')
        })

        wrapper.update()

        act(() => {
          updateTextInput(wrapper, 'lastName', 'Doh')
        })

        wrapper.update()

        act(() => {
          updateTextInput(wrapper, 'DOB', 'date')
        })

        wrapper.update()

        await act(async () => {
          await wrapper.find('#AddStudentForm').get(0).props.onSubmit({
            preventDefault: jest.fn(),
          })
          await delay()
        })

        wrapper.update()

        expect(props.toggleModal).toHaveBeenCalledTimes(1)
      })
    })
  })
})
