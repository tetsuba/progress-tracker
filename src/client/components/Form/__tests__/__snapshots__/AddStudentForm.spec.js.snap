// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddStudentForm> @Render should render 1`] = `
<AddStudentForm>
  <Form
    id="AddStudentForm"
    inline={false}
    onSubmit={[Function]}
  >
    <form
      className=""
      id="AddStudentForm"
      onSubmit={[Function]}
    >
      <FormRow
        className="pt-3"
      >
        <div
          className="pt-3 form-row"
        >
          <FormGroup
            as={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "displayName": "Col",
                "render": [Function],
              }
            }
            controlId="formGridFirstName"
          >
            <Col
              className="form-group"
            >
              <div
                className="form-group col"
              >
                <FormControl
                  name="firstName"
                  onChange={[Function]}
                  placeholder="First Name"
                  required={true}
                  size="lg"
                  value=""
                >
                  <input
                    className="form-control form-control-lg"
                    id="formGridFirstName"
                    name="firstName"
                    onChange={[Function]}
                    placeholder="First Name"
                    required={true}
                    value=""
                  />
                </FormControl>
              </div>
            </Col>
          </FormGroup>
        </div>
      </FormRow>
      <FormRow
        className="pt-3"
      >
        <div
          className="pt-3 form-row"
        >
          <FormGroup
            as={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "displayName": "Col",
                "render": [Function],
              }
            }
            controlId="formGridLastName"
          >
            <Col
              className="form-group"
            >
              <div
                className="form-group col"
              >
                <FormControl
                  name="lastName"
                  onChange={[Function]}
                  placeholder="Last Name"
                  required={true}
                  size="lg"
                  value=""
                >
                  <input
                    className="form-control form-control-lg"
                    id="formGridLastName"
                    name="lastName"
                    onChange={[Function]}
                    placeholder="Last Name"
                    required={true}
                    value=""
                  />
                </FormControl>
              </div>
            </Col>
          </FormGroup>
        </div>
      </FormRow>
      <FormLabel
        className="pt-3"
        column="lg"
        srOnly={false}
      >
        <Col
          as="label"
          className="pt-3 form-label col-form-label-lg"
        >
          <label
            className="pt-3 form-label col-form-label-lg col"
          >
            Date of birth
          </label>
        </Col>
      </FormLabel>
      <DobInput
        inputs={
          Object {
            "DOBDate": "01",
            "DOBMonth": "01",
            "DOBYear": 2020,
            "firstName": "",
            "lastName": "",
          }
        }
        setInputs={[Function]}
      >
        <FormRow>
          <div
            className="form-row"
          >
            <FormGroup
              as={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "displayName": "Col",
                  "render": [Function],
                }
              }
              controlId="formGridDOB"
            >
              <Col
                className="form-group"
              >
                <div
                  className="form-group col"
                >
                  <FormControl
                    as="select"
                    custom={true}
                    name="DOBDate"
                    onChange={[Function]}
                    size="lg"
                    value="01"
                  >
                    <select
                      className="custom-select custom-select-lg"
                      id="formGridDOB"
                      name="DOBDate"
                      onChange={[Function]}
                      value="01"
                    >
                      <option
                        disabled={true}
                      >
                        Date:
                      </option>
                      <option
                        key="dobdate-01"
                      >
                        01
                      </option>
                      <option
                        key="dobdate-02"
                      >
                        02
                      </option>
                      <option
                        key="dobdate-03"
                      >
                        03
                      </option>
                      <option
                        key="dobdate-04"
                      >
                        04
                      </option>
                      <option
                        key="dobdate-05"
                      >
                        05
                      </option>
                      <option
                        key="dobdate-06"
                      >
                        06
                      </option>
                      <option
                        key="dobdate-07"
                      >
                        07
                      </option>
                      <option
                        key="dobdate-08"
                      >
                        08
                      </option>
                      <option
                        key="dobdate-09"
                      >
                        09
                      </option>
                      <option
                        key="dobdate-10"
                      >
                        10
                      </option>
                      <option
                        key="dobdate-11"
                      >
                        11
                      </option>
                      <option
                        key="dobdate-12"
                      >
                        12
                      </option>
                      <option
                        key="dobdate-13"
                      >
                        13
                      </option>
                      <option
                        key="dobdate-14"
                      >
                        14
                      </option>
                      <option
                        key="dobdate-15"
                      >
                        15
                      </option>
                      <option
                        key="dobdate-16"
                      >
                        16
                      </option>
                      <option
                        key="dobdate-17"
                      >
                        17
                      </option>
                      <option
                        key="dobdate-18"
                      >
                        18
                      </option>
                      <option
                        key="dobdate-19"
                      >
                        19
                      </option>
                      <option
                        key="dobdate-20"
                      >
                        20
                      </option>
                      <option
                        key="dobdate-21"
                      >
                        21
                      </option>
                      <option
                        key="dobdate-22"
                      >
                        22
                      </option>
                      <option
                        key="dobdate-23"
                      >
                        23
                      </option>
                      <option
                        key="dobdate-24"
                      >
                        24
                      </option>
                      <option
                        key="dobdate-25"
                      >
                        25
                      </option>
                      <option
                        key="dobdate-26"
                      >
                        26
                      </option>
                      <option
                        key="dobdate-27"
                      >
                        27
                      </option>
                      <option
                        key="dobdate-28"
                      >
                        28
                      </option>
                      <option
                        key="dobdate-29"
                      >
                        29
                      </option>
                      <option
                        key="dobdate-30"
                      >
                        30
                      </option>
                      <option
                        key="dobdate-31"
                      >
                        31
                      </option>
                    </select>
                  </FormControl>
                </div>
              </Col>
            </FormGroup>
            <h3
              className="text-muted p-1"
            >
              /
            </h3>
            <FormGroup
              as={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "displayName": "Col",
                  "render": [Function],
                }
              }
              controlId="formGridDOB"
            >
              <Col
                className="form-group"
              >
                <div
                  className="form-group col"
                >
                  <FormControl
                    as="select"
                    custom={true}
                    name="DOBMonth"
                    onChange={[Function]}
                    size="lg"
                    value="01"
                  >
                    <select
                      className="custom-select custom-select-lg"
                      id="formGridDOB"
                      name="DOBMonth"
                      onChange={[Function]}
                      value="01"
                    >
                      <option
                        disabled={true}
                      >
                        Month:
                      </option>
                      <option>
                        01
                      </option>
                      <option>
                        02
                      </option>
                      <option>
                        03
                      </option>
                      <option>
                        04
                      </option>
                      <option>
                        05
                      </option>
                      <option>
                        06
                      </option>
                      <option>
                        07
                      </option>
                      <option>
                        08
                      </option>
                      <option>
                        09
                      </option>
                      <option>
                        10
                      </option>
                      <option>
                        11
                      </option>
                      <option>
                        12
                      </option>
                    </select>
                  </FormControl>
                </div>
              </Col>
            </FormGroup>
            <h3
              className="text-muted p-1"
            >
              /
            </h3>
            <FormGroup
              as={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "displayName": "Col",
                  "render": [Function],
                }
              }
              controlId="formGridDOB"
            >
              <Col
                className="form-group"
              >
                <div
                  className="form-group col"
                >
                  <FormControl
                    as="select"
                    custom={true}
                    name="DOBYear"
                    onChange={[Function]}
                    size="lg"
                    value={2020}
                  >
                    <select
                      className="custom-select custom-select-lg"
                      id="formGridDOB"
                      name="DOBYear"
                      onChange={[Function]}
                      value={2020}
                    >
                      <option
                        disabled={true}
                      >
                        Year:
                      </option>
                      <option
                        key="dobyear-2020"
                      >
                        2020
                      </option>
                      <option
                        key="dobyear-2019"
                      >
                        2019
                      </option>
                      <option
                        key="dobyear-2018"
                      >
                        2018
                      </option>
                      <option
                        key="dobyear-2017"
                      >
                        2017
                      </option>
                      <option
                        key="dobyear-2016"
                      >
                        2016
                      </option>
                      <option
                        key="dobyear-2015"
                      >
                        2015
                      </option>
                      <option
                        key="dobyear-2014"
                      >
                        2014
                      </option>
                      <option
                        key="dobyear-2013"
                      >
                        2013
                      </option>
                      <option
                        key="dobyear-2012"
                      >
                        2012
                      </option>
                      <option
                        key="dobyear-2011"
                      >
                        2011
                      </option>
                      <option
                        key="dobyear-2010"
                      >
                        2010
                      </option>
                      <option
                        key="dobyear-2009"
                      >
                        2009
                      </option>
                      <option
                        key="dobyear-2008"
                      >
                        2008
                      </option>
                      <option
                        key="dobyear-2007"
                      >
                        2007
                      </option>
                      <option
                        key="dobyear-2006"
                      >
                        2006
                      </option>
                      <option
                        key="dobyear-2005"
                      >
                        2005
                      </option>
                    </select>
                  </FormControl>
                </div>
              </Col>
            </FormGroup>
          </div>
        </FormRow>
      </DobInput>
      <FormRow>
        <div
          className="form-row"
        >
          <Button
            active={false}
            className="float-right"
            disabled={false}
            type="button"
            variant="secondary"
          >
            <button
              className="float-right btn btn-secondary"
              disabled={false}
              type="button"
            >
              Close
            </button>
          </Button>
          <Button
            active={false}
            className="float-right ml-2"
            disabled={false}
            type="submit"
            variant="primary"
          >
            <button
              className="float-right ml-2 btn btn-primary"
              disabled={false}
              type="submit"
            >
              Submit
            </button>
          </Button>
        </div>
      </FormRow>
    </form>
  </Form>
</AddStudentForm>
`;
